### PR TITLE
Implemented Karafka Responder for Yacs-core

### PR DESCRIPTION
**Pull request by bdieu178**
This pull request resolves the issue referenced in #285 for sending change events from YACS-core to a Kafka topic. A Kafka producer from YACS-core can now talk to a consumer.

Next step is testing the responder. @Bad-Science

### CODE REVIEW

**Comment by bdieu178** [comment]
Output of repsonder:
[2018-06-04 00:51:38,253] WARN Client session timed out, have not heard from server in 13637ms for sessionid 0x163c81a1424000c (org.apache.zookeeper.ClientCnxn)
[2018-06-04 00:51:41,916] WARN Unable to reconnect to ZooKeeper service, session 0x163c81a1424000c has expired (org.apache.zookeeper.ClientCnxn)
{"seats":5,"course_id":653,"name":"T02","crn":55515,"seats_taken":8,"created_at":"2018-06-04T00:05:43.541Z","updated_at":"2018-06-04T00:52:17.214Z","num_periods":0,"instructors":[],"conflicts":[],"uuid":"0f4b9b33-c9d4-4329-880a-e0120cd6a7b1"}
{"name":"debian","created_at":"2018-06-04T00:03:28.194Z","updated_at":"2018-06-04T00:55:13.201Z","uuid":"fb766950-692d-47fd-8ccd-e76d43cd950c"}

File: app/models/course.rb [hunk 1]
```diff
@@ -3,6 +3,13 @@ class Course < ActiveRecord::Base
   has_many   :sections, dependent: :destroy
   validates  :number, presence: true, uniqueness: { scope: :department_id }
   default_scope { order(number: :asc) }
+  after_save :respond_course
+
+
+  def respond_course
+     # CoreProducer.new.call(course)
```
Comment (Bad-Science) [inline app/models/course.rb]: remove commented line

Comment (Bad-Science) [inline app/models/course.rb]: remove commented line

File: app/models/course.rb [hunk 2]
```diff
@@ -3,6 +3,13 @@ class Course < ActiveRecord::Base
   has_many   :sections, dependent: :destroy
   validates  :number, presence: true, uniqueness: { scope: :department_id }
   default_scope { order(number: :asc) }
+  after_save :respond_course
+
+
```
Comment (Bad-Science) [inline app/models/course.rb]: remove extra newline

File: app/models/course.rb [hunk 3]
```diff
@@ -3,6 +3,13 @@ class Course < ActiveRecord::Base
   has_many   :sections, dependent: :destroy
   validates  :number, presence: true, uniqueness: { scope: :department_id }
   default_scope { order(number: :asc) }
+  after_save :respond_course
```
Comment (Bad-Science) [inline app/models/course.rb]: can we come up with a more descriptive method name that `respond_course`? That doesn't really have meaning in the context of this class.

File: app/models/section.rb [hunk 4]
```diff
@@ -5,6 +5,12 @@ class Section < ActiveRecord::Base
   default_scope { order(name: :asc) }
   before_save :sort_periods, if: :periods_changed?
   after_save :update_conflicts!, if: :periods_changed?
+  after_save :respond_section
+
+  def respond_section
```
Comment (Bad-Science) [inline app/models/section.rb]: same as `respond_course`, can we come up with a better name?

File: app/responders/core_responder.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+class CoreResponder < ApplicationResponder
+    #ALLOWED_TYPES = %w(school department course section).freeze
+
+   topic :course_change
+   topic :section_change
+
+    def respond(event)
+       respond_to :section_change , event.to_json(:only => [:uuid,:course_id, :name, :crn, :seats, :seats, :seats_taken, :created_at, :updated_at, :num_periods, :instructors, :conflicts ])
```
Comment (Bad-Science) [inline app/responders/core_responder.rb]: So it looks like right here we are writing a message to both the `section_change` and `course_change` topics. We want to only write to the appropriate topic depending on whether it is a course or a section. I think the right way to do this according to the karafka best practices would be to create two separate classes, `CoursesResponder` and `SectionsResponder`, and call them accordingly.

File: config/application.rb [hunk 6]
```diff
@@ -14,13 +14,14 @@ class Application < Rails::Application
     # Settings in config/environments/* take precedence over those specified here.
     # Application configuration should go into files in config/initializers
     # -- all .rb files in that directory are automatically loaded.
-    
+
     # Autoload all lib files
     config.autoload_paths << Rails.root.join('lib')
+    config.autoload_paths << Rails.root.join('app/responders')
```
Comment (Bad-Science) [inline config/application.rb]: I don't think we ended up needing this, I think we can delete this line

File: db/schema.rb [hunk 7]
```diff
@@ -77,7 +77,7 @@
     t.string "instructors", default: [], null: false, array: true
     t.integer "conflicts", default: [], null: false, array: true
     t.uuid "uuid", null: false
-    t.string "periods_location", default: [], array: true
+    t.string "periods_location", default: [], null: false, array: true
```
Comment (Bad-Science) [inline db/schema.rb]: not sure how this ended up in the diff

Comment (bdieu178) [reply db/schema.rb]: So would line 80 need to revert to original version?

File: app/models/section.rb [hunk 8]
```diff
@@ -5,6 +5,12 @@ class Section < ActiveRecord::Base
   default_scope { order(name: :asc) }
   before_save :sort_periods, if: :periods_changed?
   after_save :update_conflicts!, if: :periods_changed?
+  after_save :send_notification
+
+  def send_notification
+     EventResponders::SectionResponder.new.call(self)
```
Comment (Bad-Science) [inline app/models/section.rb]: use two space indentation

File: app/responders/core_responder.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
+module EventResponders
+
+    class CourseResponder < ApplicationResponder
```
Comment (Bad-Science) [inline app/responders/core_responder.rb]: use two space indentation, move each class to its own file

File: db/schema.rb (hunk 7 above)
Comment (Bad-Science) [reply db/schema.rb]: yes

File: app/responders/core_responder.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+       topic :course_changee, required: true
```
Comment (Bad-Science) [inline app/responders/core_responder.rb]: typo

File: spec/kafka/producers/course_responder_spec.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
+require "rails_helper"
```
Comment (Bad-Science) [inline spec/kafka/producers/course_responder_spec.rb]: file path should be `spec/responders/courses_spec.rb` and  `spec/responders/sections_spec.rb`

File: app/responders/core_responder.rb [hunk 12]
```diff
@@ -0,0 +1,18 @@
+module EventResponders
```
Comment (Bad-Science) [inline app/responders/core_responder.rb]: I don't think we need this module.

File: app/responders/core_responder.rb (hunk 9 above)
Comment (Bad-Science) [inline app/responders/core_responder.rb]: class names should be plural here, e.g. `CoursesResponder`

File: app/responders/core_responder.rb (hunk 12 above)
Comment (bdieu178) [reply app/responders/core_responder.rb]: Fixed

File: app/responders/core_responder.rb (hunk 9 above)
Comment (bdieu178) [reply app/responders/core_responder.rb]: Fixed

File: app/models/section.rb (hunk 8 above)
Comment (bdieu178) [reply app/models/section.rb]: Fixed

File: app/responders/core_responder.rb (hunk 9 above)
Comment (bdieu178) [reply app/responders/core_responder.rb]: Fixed

File: spec/kafka/producers/course_responder_spec.rb (hunk 11 above)
Comment (bdieu178) [reply spec/kafka/producers/course_responder_spec.rb]: Fixed

File: spec/models/section_spec.rb [hunk 13]
```diff
@@ -1,17 +1,33 @@
...
-  context 'when a section is updated' do
-    context 'when period information is updated' do
+    context 'when a section is updated' do
+      context 'notify Kafka topic' do
+         let(:param) { rand }
+         subject(:responder) { ::SectionsResponder.new }
+
+           describe '#send_notification' do
+                 it 'executes call method with self parameter' do
+                     expect(responder.call(self))
```
Comment (Bad-Science) [inline spec/models/section_spec.rb]: I don't think this is actually testing anything. it looks like this is just expecting the return value of `responder.call(self)` to be truthy. See: https://stackoverflow.com/questions/16677718/how-to-test-models-callback-method-independently

File: spec/models/section_spec.rb [hunk 14]
```diff
@@ -1,17 +1,33 @@
...
+                 end
+
+                 it 'cannot accept another parameter' do
```
Comment (Bad-Science) [inline spec/models/section_spec.rb]: Not sure exactly why we need this, but it looks like this should go in the responder test class.

File: app/models/section.rb [hunk 15]
```diff
@@ -5,6 +5,13 @@ class Section < ActiveRecord::Base
   default_scope { order(name: :asc) }
   before_save :sort_periods, if: :periods_changed?
   after_save :update_conflicts!, if: :periods_changed?
+  after_save :send_notification
+
+
```
[the remark this answers is not in the document]
Comment (bdieu178) [reply app/models/section.rb]: From your prior review you asked me to add this additional line.

File: app/models/course.rb [hunk 16]
```diff
@@ -3,6 +3,12 @@ class Course < ActiveRecord::Base
   has_many   :sections, dependent: :destroy
   validates  :number, presence: true, uniqueness: { scope: :department_id }
   default_scope { order(number: :asc) }
+  after_save :send_notification
+
+
+  def send_notification
+    ::CoursesResponder.new.call(self)
```
[the remark this answers is not in the document]
Comment (Bad-Science) [reply app/models/course.rb]: From my recollection we determined that was not the cause. If removing them still causes errors though then keep them :P

File: app/models/section.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (Bad-Science) [reply app/models/section.rb]: https://github.com/YACS-RCOS/yacs/pull/287#discussion_r192934432

[the remark this answers is not in the document]
Comment (bdieu178) [reply app/models/section.rb]: I confused indentation with newline lmao

File: spec/models/section_spec.rb (hunk 13 above)
Comment (bdieu178) [reply spec/models/section_spec.rb]: In general for unit tests do we add or do not add a test that outputs a value we expect? I just want to understand the best way to test moving forward.

Comment (Bad-Science) [reply spec/models/section_spec.rb]: It depends. If we are testing an individual method (which is often the case), then yes. That's not the case here though. Here we want to test that when a certain action (updating a section) occurs, a certain method (`SectionsResponder#call`) is called. Such tests are generally less desirable, but are often necessary when testing event-driven behavior that heavily intertwined with a framework.

**Comment by bdieu178** [comment]
@Bad-Science  Please review this section of the docker-compose.yml file to check whether I implemented the environment variables properly for the Karafka responders.

screen shot 2018-06-11 at 8 36 54 pm

File: app/responders/sections_responder.rb [hunk 17]
```diff
@@ -0,0 +1,9 @@
+class SectionsResponder < ApplicationResponder
+  @@sections_topic_name = ENV['SECTIONS_TOPIC_NAME']
```
Comment (Bad-Science) [inline app/responders/sections_responder.rb]: use https://apidock.com/ruby/String/to_sym and remove interpolation

`@@topic_name = ENV['SECTIONS_TOPIC_NAME'].to_sym`

File: spec/models/course_spec.rb [hunk 18]
```diff
@@ -5,6 +5,21 @@
...
     end
+
+    context 'when a course is updated' do
+      context 'notify Kafka topic' do
+        let(:param) { rand }
+        subject(:responder) { CoursesResponder.new }
+
+        describe '#send_notification' do
+          it 'executes call method with self parameter' do
+            expect(@course).to receive(send_notification)
```
Comment (Bad-Science) [inline spec/models/course_spec.rb]: Does this test pass? If it does then there is something wrong. The `send_notification` method should be called iff the section is changed, and it's not being changed here.

File: spec/responders/coursesresponder_spec.rb [hunk 19]
```diff
@@ -0,0 +1,28 @@
+require "rails_helper"
```
Comment (Bad-Science) [inline spec/responders/coursesresponder_spec.rb]: see previous comment, filename should be `courses_spec.rb`

File: spec/responders/sectionsresponder_spec.rb [hunk 20]
```diff
@@ -0,0 +1,27 @@
+require "rails_helper"
```
Comment (Bad-Science) [inline spec/responders/sectionsresponder_spec.rb]: see previous comment, filename should be `sections_spec.rb`

File: app/responders/sections_responder.rb (hunk 17 above)
Comment (bdieu178) [reply app/responders/sections_responder.rb]: I've made the change. Nevertheless, can you please clarify by what you mean by interpolation in this context, please?

Comment (Bad-Science) [reply app/responders/sections_responder.rb]: http://ruby-for-beginners.rubymonstas.org/bonus/string_interpolation.html

File: spec/models/course_spec.rb [hunk 21]
```diff
@@ -5,22 +5,32 @@
     before do
       @course = create(:course)
     end
-    context 'when there is a section' do
+
+    context 'when a course is updated' do
+      describe '#send_notification' do
+        it 'executes call method with self parameter' do
+          @section.save
```
Comment (Bad-Science) [inline spec/models/course_spec.rb]: So, calling `#save` actually won't trigger an update unless a field is actually changed. I'm going to merge this anyway though, and we can go back and do a pass over the tests before merging this into staging.

**Pull request merged**